### PR TITLE
fog w shadows & ssao

### PR DESCRIPTION
**Pull request by dmorehead**
simplified "fog + shadows" to no longer have to use a custom GL_BLEND or use gl.drawArrays twice within the shadow pass. Instead use fog to "mask" the shadows so only the "shadows - fog" are returned from renderShadows. Also applied this same logic to the SSAO pass as it feels more natural visually to have the SSAO disappear into the fog.
fogShadSSAO

### CODE REVIEW

**Comment by andreasplesch** [comment]
Thanks ! I think this could work well as simplifications are always welcome. Please run `npm run lint-fix` which should take care of the formatting. There may also be some comments on the code.

File: src/SSAO.js [hunk 1]
```diff
@@ -255,6 +255,12 @@ x3dom.SSAO.blur = function ( stateManager, gl, scene, ssaoTexture, depthTexture,
 
     stateManager.useProgram( sp );
 
+    // Mask ssao w/ fog, no need for fogColor
+    // ALWAYS pass FOG uniforms into shader or scenes w/out fog will barf
```
Comment (andreasplesch) [inline src/SSAO.js:259]: Could you check for fog here and in the SSAO shader (properties.FOG, see below) ?

File: src/shader/ShaderDynamic.js [hunk 2]
```diff
@@ -742,19 +742,33 @@ x3dom.shader.DynamicShader.prototype.generateFragmentShader = function ( gl, pro
...
+    if ( properties.LIGHTS || properties.FOG || properties.CLIPPLANES )
     {
+        shader += "uniform vec3 eyePosition;\n";
         shader += "varying vec4 fragPosition;\n";
-        shader += "varying vec4 fragPositionWS;\n";
         shader += "uniform float isOrthoView;\n";
-    }
+        shader += "varying vec4 fragPositionWS;\n";
 
+        if ( properties.FOG )
```
Comment (andreasplesch) [inline src/shader/ShaderDynamic.js:766]: seems cleaner to keep outside and not to nest inside here.

File: src/shader/ShaderDynamic.js [hunk 3]
```diff
@@ -742,19 +742,33 @@ x3dom.shader.DynamicShader.prototype.generateFragmentShader = function ( gl, pro
...
+    // {
+    //     shader += "varying vec4 fragPosition;\n";
+    //     shader += "varying vec4 fragPositionWS;\n";
+    //     shader += "uniform float isOrthoView;\n";
+    // }
+
+    // same as vertex shader but with fragPositionWS for fogNoise (w/ or w/out lights)
+    if ( properties.LIGHTS || properties.FOG || properties.CLIPPLANES )
     {
+        shader += "uniform vec3 eyePosition;\n";
```
Comment (andreasplesch) [inline src/shader/ShaderDynamic.js:761]: Is this needed ?

File: src/SSAO.js (hunk 1 above)
Comment (dmorehead) [reply src/SSAO.js:259]: ideally yes, however (unlike shapes & shadows) properties are not being created in the ssao chain, and I was hesitant to shoehorn them into x3dom.Utils.wrapProgram.

File: src/shader/ShaderDynamic.js (hunk 2 above)
Comment (dmorehead) [reply src/shader/ShaderDynamic.js]: originally I nested this check to mimic the exact logic used to create the corresponding vertex shader, but looks fine to moved it back outside.

File: src/shader/ShaderDynamic.js (hunk 3 above)
Comment (dmorehead) [reply src/shader/ShaderDynamic.js]: looks like it is only needed by the vertex shader, everything seems to work w/out it.

**Review by andreasplesch — APPROVED** [review]

File: src/shader/ShaderShadowRendering.js [hunk 4]
```diff
@@ -172,15 +163,25 @@ x3dom.shader.ShadowRenderingShader.prototype.generateFragmentShader = function (
...
     // gamma coefficient), i.e. the umbra is corrected for now, the penumbra
     // is incorrect and full light is zero here so unaffected as well.
+    shader += "    color = " + x3dom.shader.encodeGamma( {}, "vec4(shadowValue, shadowValue, shadowValue, 1.0)" ) + " ;\n";
+
+    if ( properties.FOG )
+    {
+    // mask shadow w/ fogDepth, visually no need for fogColor
+        shader += "   vec3 fragEyePosition = eyeCoords.xyz / eyeCoords.w;\n" +
+              "   float f0 = calcFog(fragEyePosition);\n" +
+        //"   color.rgb = fogColor * (1.0 - f0) + f0 * color.rgb;\n";
```
Comment (andreasplesch) [inline src/shader/ShaderShadowRendering.js:178]: Please remove comment.

**Review by andreasplesch — CHANGES_REQUESTED** [review]
Please add a line to CHANGELOG.md

File: src/shader/ShaderShadowRendering.js (hunk 4 above)
Comment (andreasplesch) [reply src/shader/ShaderShadowRendering.js]: Sorry, I meant the comment in line 178.

**Review by andreasplesch — APPROVED** [review]

**Pull request merged**